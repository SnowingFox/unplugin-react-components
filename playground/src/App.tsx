import './App.css'

function App() {
  return (
    <div className="App">
      <Button variant={'contained'}>hi mui</Button>
      <AntProgress percent={30} />
      <AntSkeleton />
      <AntTooltip title="prompt text">
        <span>Tooltip will show on mouse enter.</span>
      </AntTooltip>
      <CompC />
    </div>
  )
}

export default App
