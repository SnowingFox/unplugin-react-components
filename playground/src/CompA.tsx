export const CompA: React.FC = () => {
  return (
    <div>CompA</div>
  )
}
