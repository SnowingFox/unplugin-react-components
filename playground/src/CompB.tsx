export const CompB = () => { 
  return (
    <div>CompB</div>
  )
}
