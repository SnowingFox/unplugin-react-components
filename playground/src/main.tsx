import ReactDOM from 'react-dom/client'
import React from 'react'
import './index.css'

ReactDOM.createRoot(document.getElementById('root') as HTMLElement).render(
  <React.StrictMode>
    <App />
  </React.StrictMode>,
)
