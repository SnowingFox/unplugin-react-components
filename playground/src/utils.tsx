export const sum = (a: number, b: number) => a + b

export function AAA() { return <div>1</div> }
export default function Utils() { return <div>1</div> }

