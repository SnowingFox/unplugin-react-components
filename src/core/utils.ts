import type { ComponentsContext, ExportType, ImportInfo, Options } from '../types'

export const isExportComponent = (component: ComponentsContext | ExportType) =>
  typeof component === 'string' ? component === 'Export' : component.type === 'Export'

export const isCapitalCase = (code: string) => {
  const ascii = code[0].charCodeAt(0)
  return ascii >= 65 && ascii <= 90
}
/**
 * Replace backslash to slash
 */
export function slash(str: string) {
  return str.replace(/\\/g, '/')
}

export function stringifyImport(info: ImportInfo | string) {
  if (typeof info === 'string')
    return `import '${info}'`
  else if (info.name && info.as)
    return `import { ${info.name} as ${info.as} } from '${info.from}'`
  else if (info.name)
    return `import { ${info.name} } from '${info.from}'`
  else
    return `import ${info.default} from '${info.from}'`
}

export function resolveOptions(options: Options = {}): Required<Options> {
  return {
    rootDir: options.rootDir || process.cwd(),
    dts: false,
    include: [/\.[j|t]sx$/],
    exclude: [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/],
    resolvers: options.resolvers || [],
    local: typeof options.local === 'boolean' ? options.local : true,
    mode: 'dev',
    ...options,
  }
}
