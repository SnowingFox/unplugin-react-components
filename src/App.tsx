import { useState } from 'react'
import './App.css'
import { CompA } from './CompA'
import CompB from './CompB'

type Com = typeof import('./CompA').CompA

function App() {
  const [count, setCount] = useState(0)

  return (
    <div className="App">
      <CompA />
      <CompB />
    </div>
  )
}

export default App
