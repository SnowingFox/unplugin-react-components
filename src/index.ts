import { createUnplugin } from 'unplugin'
import MagicString from 'magic-string'
import { createFilter } from '@rollup/pluginutils'
import type { GenerateDtsOptions, Options, TransformOptions } from './types'
import { changeREOnBuildStart, transform } from './core/transformer'
import { searchGlob } from './core/searchGlob'
import { generateDts } from './core/generateDts'
import { resolveOptions } from './core/utils'

export * from './core/resolvers'
export * from './core/generateDts'
export * from './core/resolvers/index'
export * from './core/searchGlob'
export * from './core/transformer'

export default createUnplugin<Options>((options = {}) => {
  options = resolveOptions(options)

  const filter = createFilter(
    options.include || [/\.[j|t]sx$/],
    options.exclude || [/[\\/]node_modules[\\/]/, /[\\/]\.git[\\/]/],
  )
  const searchGlobResult = searchGlob({ rootPath: (options.dts as any)?.rootPath || options.rootDir! })
  const dtsOptions = {
    components: searchGlobResult,
    filename: (options.dts as GenerateDtsOptions)?.filename || 'components',
    rootPath: (options.dts as GenerateDtsOptions)?.rootPath || options.rootDir!,
    local: options.local,
    resolvers: options.resolvers,
  } as GenerateDtsOptions

  if (options.dts === true)
    generateDts({ ...dtsOptions })
  else if (typeof options.dts === 'object')
    generateDts({ ...dtsOptions, ...options.dts })

  return {
    name: 'unplugin-react-components',
    transformInclude(id) {
      return filter(id)
    },
    async transform(code, id) {
      const context: TransformOptions = {
        code: new MagicString(code),
        components: searchGlobResult,
        rootDir: options.rootDir!,
        resolvers: options.resolvers!,
        local: options.local!,
        mode: options.mode!,
        id,
      }

      return await transform(context)
    },
    buildStart() {
      if (options.mode === 'prod')
        changeREOnBuildStart()
    },
  }
})
