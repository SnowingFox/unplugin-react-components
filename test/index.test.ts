import { resolve } from 'path'
import { expect, test } from 'vitest'
import MagicString from 'magic-string'
import { AntdResolver, generateDts, searchGlob, transform } from '../src'
import { slash } from '../src/core/utils'

const rootPath = slash(`${resolve(__dirname)}/fixtures`)
const searchGlobResult = searchGlob({
  rootPath,
})

const id = slash(`${process.cwd()}/test/fixtures/App.tsx`)

const code = `
  import { jsxDEV as _jsxDEV } from "react/jsx-dev-runtime";
  function App() {
    return /*#__PURE__*/_jsxDEV("div", {
      className: "App",
      children: [/*#__PURE__*/_jsxDEV(A, {
        variant: 'contained',
        children: "hi mui"
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 6,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(AntProgress, {
        percent: 30
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 7,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(A, {}, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 8,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(AntTooltip, {
        title: "prompt text",
        children: /*#__PURE__*/_jsxDEV("span", {
          children: "Tooltip will show on mouse enter."
        }, void 0, false, {
          fileName: _jsxFileName,
          lineNumber: 10,
          columnNumber: 9
        }, this)
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 9,
        columnNumber: 7
      }, this)]
    }, void 0, true, {
      fileName: _jsxFileName,
      lineNumber: 5,
      columnNumber: 5
    }, this);
  }
  _c = App;
  export default App;`

test('test searchGlob', async () => {
  const arr = Array.from(searchGlobResult).map(i => ({ ...i, path: 'searchGlob' }))
  expect(arr).toMatchInlineSnapshot(`
    [
      {
        "name": "A",
        "path": "searchGlob",
        "type": "Export",
      },
      {
        "name": "App",
        "path": "searchGlob",
        "type": "Export",
      },
    ]
  `)
})

test('play with resolver', async () => {
  const transformed = await transform({
    code: new MagicString(code),
    id,
    components: searchGlobResult,
    resolvers: [
      AntdResolver({
        prefix: 'Ant',
      }),
    ],
    local: true,
    rootDir: '',
  })

  transformed.replace('""', '"')

  expect(transformed).toMatchSnapshot()
})

test('ignore local components', async () => {
  expect(await transform({
    code: new MagicString(code),
    id,
    components: searchGlobResult,
    local: false,
    rootDir: slash(`${resolve(__dirname)}/fixtures`),
    resolvers: [
      AntdResolver(),
    ],
  })).toMatchSnapshot()
})

test('generate components.d.ts', async () => {
  const dts = await generateDts({
    components: searchGlobResult,
    filename: 'components',
    local: true,
    rootPath,
    resolvers: [],
  })

  expect(dts).toMatchSnapshot()
})

test('it should work after bundle.', async () => {
  const code = `import "./App.css";
import { jsx as _jsx } from "react/jsx-runtime";
import { jsxs as _jsxs } from "react/jsx-runtime";
function App() {
  return /* @__PURE__ */ _jsxs("div", {
    className: "App",
    children: [/* @__PURE__ */ _jsx(Button, {
      variant: "contained",
      children: "hi mui"
    }), /* @__PURE__ */ _jsx(Box, {}), /* @__PURE__ */ _jsx(AntProgress, {
      percent: 30
    }), /* @__PURE__ */ _jsx(AntProgress, {
      percent: 30
    }), /* @__PURE__ */ _jsx(AntSkeleton, {}), /* @__PURE__ */ _jsx(AntTooltip, {
      title: "prompt text",
      children: /* @__PURE__ */ _jsx("span", {
        children: "Tooltip will show on mouse enter."
      })
    }), /* @__PURE__ */ _jsx(CompC, {})]
  });
}
export default App;
`

  expect(await transform({
    code: new MagicString(code),
    id,
    components: searchGlobResult,
    local: false,
    rootDir: slash(`${resolve(__dirname)}/fixtures`),
    resolvers: [
      AntdResolver(),
    ],
  })).toMatchSnapshot()
})
