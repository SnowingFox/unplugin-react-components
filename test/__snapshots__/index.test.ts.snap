// Vitest Snapshot v1

exports[`generate components.d.ts 1`] = `
"/* generated by unplugin-react-components */
export {}
declare global{
	const A: typeof import('./A')['A']
	const App: typeof import('./App')['App']
	const AntAffix: typeof import('antd')['Affix']
	const AntAlert: typeof import('antd')['Alert']
	const AntAnchor: typeof import('antd')['Anchor']
	const AntAutoComplete: typeof import('antd')['AutoComplete']
	const AntAvatar: typeof import('antd')['Avatar']
	const AntBackTop: typeof import('antd')['BackTop']
	const AntBadge: typeof import('antd')['Badge']
	const AntBreadcrumb: typeof import('antd')['Breadcrumb']
	const AntButton: typeof import('antd')['Button']
	const AntCalendar: typeof import('antd')['Calendar']
	const AntCard: typeof import('antd')['Card']
	const AntCarousel: typeof import('antd')['Carousel']
	const AntCascader: typeof import('antd')['Cascader']
	const AntCheckbox: typeof import('antd')['Checkbox']
	const AntCol: typeof import('antd')['Col']
	const AntCollapse: typeof import('antd')['Collapse']
	const AntComment: typeof import('antd')['Comment']
	const AntConfigProvider: typeof import('antd')['ConfigProvider']
	const AntDatePicker: typeof import('antd')['DatePicker']
	const AntDescriptions: typeof import('antd')['Descriptions']
	const AntDivider: typeof import('antd')['Divider']
	const AntDrawer: typeof import('antd')['Drawer']
	const AntDropdown: typeof import('antd')['Dropdown']
	const AntEmpty: typeof import('antd')['Empty']
	const AntForm: typeof import('antd')['Form']
	const AntGrid: typeof import('antd')['Grid']
	const AntImage: typeof import('antd')['Image']
	const AntInput: typeof import('antd')['Input']
	const AntInputNumber: typeof import('antd')['InputNumber']
	const AntLayout: typeof import('antd')['Layout']
	const AntList: typeof import('antd')['List']
	const AntMentions: typeof import('antd')['Mentions']
	const AntMenu: typeof import('antd')['Menu']
	const AntModal: typeof import('antd')['Modal']
	const AntPageHeader: typeof import('antd')['PageHeader']
	const AntPagination: typeof import('antd')['Pagination']
	const AntPopconfirm: typeof import('antd')['Popconfirm']
	const AntPopover: typeof import('antd')['Popover']
	const AntProgress: typeof import('antd')['Progress']
	const AntRadio: typeof import('antd')['Radio']
	const AntRate: typeof import('antd')['Rate']
	const AntResult: typeof import('antd')['Result']
	const AntRow: typeof import('antd')['Row']
	const AntSegmented: typeof import('antd')['Segmented']
	const AntSelect: typeof import('antd')['Select']
	const AntSkeleton: typeof import('antd')['Skeleton']
	const AntSlider: typeof import('antd')['Slider']
	const AntSpace: typeof import('antd')['Space']
	const AntSpin: typeof import('antd')['Spin']
	const AntStatistic: typeof import('antd')['Statistic']
	const AntSteps: typeof import('antd')['Steps']
	const AntSwitch: typeof import('antd')['Switch']
	const AntTable: typeof import('antd')['Table']
	const AntTabs: typeof import('antd')['Tabs']
	const AntTag: typeof import('antd')['Tag']
	const AntTimePicker: typeof import('antd')['TimePicker']
	const AntTimeline: typeof import('antd')['Timeline']
	const AntTooltip: typeof import('antd')['Tooltip']
	const AntTransfer: typeof import('antd')['Transfer']
	const AntTree: typeof import('antd')['Tree']
	const AntTreeSelect: typeof import('antd')['TreeSelect']
	const AntTypography: typeof import('antd')['Typography']
	const AntUpload: typeof import('antd')['Upload']
	const AntAffix: typeof import('antd')['Affix']
	const AntAlert: typeof import('antd')['Alert']
	const AntAnchor: typeof import('antd')['Anchor']
	const AntAutoComplete: typeof import('antd')['AutoComplete']
	const AntAvatar: typeof import('antd')['Avatar']
	const AntBackTop: typeof import('antd')['BackTop']
	const AntBadge: typeof import('antd')['Badge']
	const AntBreadcrumb: typeof import('antd')['Breadcrumb']
	const AntButton: typeof import('antd')['Button']
	const AntCalendar: typeof import('antd')['Calendar']
	const AntCard: typeof import('antd')['Card']
	const AntCarousel: typeof import('antd')['Carousel']
	const AntCascader: typeof import('antd')['Cascader']
	const AntCheckbox: typeof import('antd')['Checkbox']
	const AntCol: typeof import('antd')['Col']
	const AntCollapse: typeof import('antd')['Collapse']
	const AntComment: typeof import('antd')['Comment']
	const AntConfigProvider: typeof import('antd')['ConfigProvider']
	const AntDatePicker: typeof import('antd')['DatePicker']
	const AntDescriptions: typeof import('antd')['Descriptions']
	const AntDivider: typeof import('antd')['Divider']
	const AntDrawer: typeof import('antd')['Drawer']
	const AntDropdown: typeof import('antd')['Dropdown']
	const AntEmpty: typeof import('antd')['Empty']
	const AntForm: typeof import('antd')['Form']
	const AntGrid: typeof import('antd')['Grid']
	const AntImage: typeof import('antd')['Image']
	const AntInput: typeof import('antd')['Input']
	const AntInputNumber: typeof import('antd')['InputNumber']
	const AntLayout: typeof import('antd')['Layout']
	const AntList: typeof import('antd')['List']
	const AntMentions: typeof import('antd')['Mentions']
	const AntMenu: typeof import('antd')['Menu']
	const AntModal: typeof import('antd')['Modal']
	const AntPageHeader: typeof import('antd')['PageHeader']
	const AntPagination: typeof import('antd')['Pagination']
	const AntPopconfirm: typeof import('antd')['Popconfirm']
	const AntPopover: typeof import('antd')['Popover']
	const AntProgress: typeof import('antd')['Progress']
	const AntRadio: typeof import('antd')['Radio']
	const AntRate: typeof import('antd')['Rate']
	const AntResult: typeof import('antd')['Result']
	const AntRow: typeof import('antd')['Row']
	const AntSegmented: typeof import('antd')['Segmented']
	const AntSelect: typeof import('antd')['Select']
	const AntSkeleton: typeof import('antd')['Skeleton']
	const AntSlider: typeof import('antd')['Slider']
	const AntSpace: typeof import('antd')['Space']
	const AntSpin: typeof import('antd')['Spin']
	const AntStatistic: typeof import('antd')['Statistic']
	const AntSteps: typeof import('antd')['Steps']
	const AntSwitch: typeof import('antd')['Switch']
	const AntTable: typeof import('antd')['Table']
	const AntTabs: typeof import('antd')['Tabs']
	const AntTag: typeof import('antd')['Tag']
	const AntTimePicker: typeof import('antd')['TimePicker']
	const AntTimeline: typeof import('antd')['Timeline']
	const AntTooltip: typeof import('antd')['Tooltip']
	const AntTransfer: typeof import('antd')['Transfer']
	const AntTree: typeof import('antd')['Tree']
	const AntTreeSelect: typeof import('antd')['TreeSelect']
	const AntTypography: typeof import('antd')['Typography']
	const AntUpload: typeof import('antd')['Upload']
}"
`;

exports[`ignore local components 1`] = `
"import { Progress as _unplugin_react_Progress_0 } from 'antd'
import { Tooltip as _unplugin_react_Tooltip_1 } from 'antd'

  import { jsxDEV as _jsxDEV } from \\"react/jsx-dev-runtime\\";
  function App() {
    return /*#__PURE__*/_jsxDEV(\\"div\\", {
      className: \\"App\\",
      children: [/*#__PURE__*/_jsxDEV(A, {
        variant: 'contained',
        children: \\"hi mui\\"
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 6,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(_unplugin_react_Progress_0, {
        percent: 30
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 7,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(A, {}, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 8,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(_unplugin_react_Tooltip_1, {
        title: \\"prompt text\\",
        children: /*#__PURE__*/_jsxDEV(\\"span\\", {
          children: \\"Tooltip will show on mouse enter.\\"
        }, void 0, false, {
          fileName: _jsxFileName,
          lineNumber: 10,
          columnNumber: 9
        }, this)
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 9,
        columnNumber: 7
      }, this)]
    }, void 0, true, {
      fileName: _jsxFileName,
      lineNumber: 5,
      columnNumber: 5
    }, this);
  }
  _c = App;
  export default App;"
`;

exports[`play with resolver 1`] = `
"import { A as _unplugin_react_A_0 } from 'D:/Dev/unplugin/unplugin-react-components/test/fixtures/A.tsx'
import { Progress as _unplugin_react_Progress_1 } from 'antd'
import { Tooltip as _unplugin_react_Tooltip_2 } from 'antd'

  import { jsxDEV as _jsxDEV } from \\"react/jsx-dev-runtime\\";
  function App() {
    return /*#__PURE__*/_jsxDEV(\\"div\\", {
      className: \\"App\\",
      children: [/*#__PURE__*/_jsxDEV(_unplugin_react_A_0, {
        variant: 'contained',
        children: \\"hi mui\\"
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 6,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(_unplugin_react_Progress_1, {
        percent: 30
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 7,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(_unplugin_react_A_0, {}, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 8,
        columnNumber: 7
      }, this), /*#__PURE__*/_jsxDEV(_unplugin_react_Tooltip_2, {
        title: \\"prompt text\\",
        children: /*#__PURE__*/_jsxDEV(\\"span\\", {
          children: \\"Tooltip will show on mouse enter.\\"
        }, void 0, false, {
          fileName: _jsxFileName,
          lineNumber: 10,
          columnNumber: 9
        }, this)
      }, void 0, false, {
        fileName: _jsxFileName,
        lineNumber: 9,
        columnNumber: 7
      }, this)]
    }, void 0, true, {
      fileName: _jsxFileName,
      lineNumber: 5,
      columnNumber: 5
    }, this);
  }
  _c = App;
  export default App;"
`;
